채팅 메세지 가져오기

jungyoubin opened the pull request:
### PR Checklist ⛑️

- [x] Commit Log Convention을 준수하여 커밋 메시지를 작성했습니다.
- [x] 변경사항이 정상적으로 작동하는지 테스트를 했습니다.

### 중요 변경 사항 🚨

> 하기 사항이 변경된 경우 체크해주세요.

- [ ] DB 스키마가 수정되었습니다.
- [ ] 환경변수 파일(`.env`)이 추가/삭제/수정되었습니다.
- [ ] `npm` 라이브러리가 추가/삭제/업데이트되었습니다.



### 작업 내용  🧑🏻‍💻
> 기능 추가, 버그 수정 등 작업 내용을 작성해주세요

- 특정 방에 대한 채팅 메세지를 불러옵니다
- 커서 페이지네이션을 통해 메시지들을 불러옵니다
- `uuidv7`의 타입을 가진 `messageId`를 통해 정렬 및 커서를 지정합니다

### 관련 이슈

> 본 PR이 머지되면 자동으로 닫을 이슈를 [형식](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)에 따라 작성하고,
> 종료하지 않고 연결만 한다면 예약어 없이 작성해주세요.

- close #33

jungyoubin committed:
feat: 채팅 메세지 가져오기 api 추가

jungyoubin committed:
fix: 주석 제거

Yuhan-Jeong commented on src/chat/chat.service.ts:
```diff
@@ -297,4 +297,33 @@ export class ChatService {
+      messageDate: new Date(r.messageDate).toISOString(), // ISO 문자열로 통일
+    }));
+
+    const last = rows[rows.length - 1];
+    const nextCursor = last ? new Date(last.messageDate).toISOString() : null;
```
`last`는 항상 값이 존재할 것으로 보입니다.

jungyoubin replied:
``` ts
  // 다음 커서는 이번 페이지의 '가장 과거' 시간
    const last = rows[rows.length - 1];
    const nextCursor = new Date(last.messageDate).toISOString();

    // 마지막 시간보다 과거 문서 존재 여부만 확인
    const more = await this.chatMessageModel.exists({
      roomId,
      messageDate: { $lt: last.messageDate }, // $lt: ~~ : ~~ 보다 작은 문서 조회
    });
    const hasMore = !!more; // 있으면 true, 없으면 false
```

다음으로 수정하였습니다.

`last`변수를 통해서 가져오는 메세지 중에서 가장 마지막의 메시지 선택합니다
`nextCursor`를 통해서 가장 마지막의 메시지의 날짜를 추출합니다
`more`에서는 선택된 가장 마지막의 메시지보다 더 과거의 메시지가 있는지 확인을 합니다
그래서 `hasMore`에서는 있는지의 유무를 확인합니다.

만약에 더 있을 경우, nextCursor의 값을 입력하여 더 이전의 값을 불러올 수 있게 진행하였습니다.

Yuhan-Jeong replied:
@jungyoubin 인턴님, 단순하게 다음 데이터가 있는지 확인하기 위해 MongoDB에 한번 더 접근하는 과정은 불요한 것 같습니다.

조회할 데이터의 수(`limit`)와 조회한 데이터의 수(`rows.length`)를 알고 있으니 이 둘을 비교하는 방식을 사용하는건 어떨까요?

jungyoubin replied:
해당 부분에 대해서 설명을 들었습니다. 

굳이 DB조회를 하지 않고 기존에 가지고 있는 정보들을 통해 해당 데이터들이 수를 비교하는 방식으로 수정 작업을 진행하였습니다

감사합니다.

Yuhan-Jeong commented on src/chat/chat.service.ts:
```diff
@@ -297,4 +297,33 @@ export class ChatService {
+    const nextCursor = last ? new Date(last.messageDate).toISOString() : null;
+    const hasMore = rows.length === limit;
+
+    return { roomId, messages, nextCursor, hasMore };
```
유저의 닉네임 반환은 불요한 것일까요?

jungyoubin replied:
감사합니다.

해당 부분 추가하였습니다.

Yuhan-Jeong commented on src/chat/chat.controller.ts:
```diff
@@ -67,4 +68,18 @@ export class ChatController {
+    @Query('limit') limits: string,
+    @Query('cursor') cursorQ: string,
+  ) {
+    const parsed = parseInt(limits ?? '50', 10);
+    const limit = isNaN(parsed) ? 50 : parsed;
+    const cursor = cursorQ ? new Date(cursorQ) : undefined; // 없으면 undefined
```
QueryParams의 타입때문이라면 Dto를 활용하는 방안은 어떨까요? 의견 부탁드립니다.

jungyoubin replied:
좋은 의견 감사합니다. 

DTO 에서 관리하는게 더 편리한 것 같습니다.

Yuhan-Jeong reviewed (changes requested):
수고하셨습니다. 리뷰 확인 부탁드립니다.

sungmin22 commented on src/chat/chat.controller.ts:
```diff
@@ -67,4 +68,18 @@ export class ChatController {
     return this.chatService.leaveRoom(user.sub, roomId);
   }
+
+  @UseGuards(JwtAuthGuard)
+  @Get('room/:roomId/message')
+  async listRoomMessages(
+    @Param('roomId') roomId: string,
+    @Query('limit') limits: string,
+    @Query('cursor') cursorQ: string,
```
- API Path와 쿼리파라미터에 대한 내용이 명세와 상이한 것으로 보입니다. 확인 부탁드립니다.
- 쿼리파라미터 키값과 변수명을 동일하게 가져가는게 가독성이 좋고 로직에서 혼선이 없을 것 같습니다.

jungyoubin replied:
넵 감사합니다.

sungmin22 commented on src/chat/chat.controller.ts:
```diff
@@ -67,4 +68,18 @@ export class ChatController {
+    @Query('limit') limits: string,
+    @Query('cursor') cursorQ: string,
+  ) {
+    const parsed = parseInt(limits ?? '50', 10);
+    const limit = isNaN(parsed) ? 50 : parsed;
```
쿼리파라미터로 받는 `limits` 혹은 `getRoomMessages()` 변수 중 `limit`에 기본값을 50으로 설정해주면 해당 로직 없이 사용 가능 할 것으로 보이는데, 의견 부탁드립니다.

jungyoubin replied:
해당 부분에 대해서 DTO 로 따로 빼서 진행하였습니다.

sungmin22 commented on src/chat/chat.service.ts:
```diff
@@ -297,4 +297,33 @@ export class ChatService {
+      messageDate: new Date(r.messageDate).toISOString(), // ISO 문자열로 통일
+    }));
+
+    const last = rows[rows.length - 1];
+    const nextCursor = last ? new Date(last.messageDate).toISOString() : null;
+    const hasMore = rows.length === limit;
```
- `last.messageDate`에 대해 불요한 ISO 문자열 변환 로직이 중복으로 들어간 것 같습니다. 확인 부탁드립니다.

- `hasMore`가 별도로 필요한 이유가 있을까요? 불러와야하는 메시지가 더 존재하는지는 확인은 `nextCursor` 존재여부로 확인 가능해보여 문의드립니다.

jungyoubin replied:
``` ts
  // 다음 커서는 이번 페이지의 '가장 과거' 시간
    const last = rows[rows.length - 1];
    const nextCursor = new Date(last.messageDate).toISOString();

    // 마지막 시간보다 과거 문서 존재 여부만 확인
    const more = await this.chatMessageModel.exists({
      roomId,
      messageDate: { $lt: last.messageDate }, // $lt: ~~ : ~~ 보다 작은 문서 조회
    });
    const hasMore = !!more; // 있으면 true, 없으면 false
```

다음으로 수정하였습니다.

`last`변수를 통해서 가져오는 메세지 중에서 가장 마지막의 메시지 선택합니다
`nextCursor`를 통해서 가장 마지막의 메시지의 날짜를 추출합니다
`more`에서는 선택된 가장 마지막의 메시지보다 더 과거의 메시지가 있는지 확인을 합니다
그래서 `hasMore`에서는 있는지의 유무를 확인합니다.

만약에 더 있을 경우, nextCursor의 값을 입력하여 더 이전의 값을 불러올 수 있게 진행하였습니다.

jungyoubin replied:
해당 중복으로 들어간 부분을 제거 하였으며, 더 남아있는 여부에 대해서 DB 조회가 아닌 문자열 길이를 비교하는 기능으로 수정 작업을 진행하였습니다.

sungmin22 reviewed (commented):
수고하셨습니다! 리뷰 확인 부탁드립니다.

jungyoubin committed:
Merge branch 'main' into 33-get-old-message

jungyoubin committed:
Merge branch 'main' into 33-get-old-message

jungyoubin committed:
Update chat.controller.ts

jungyoubin committed:
feat: 메시지 가져오는 방법에 대한 로직 수정

jungyoubin committed:
feat: 메시지를 가져올때, 사용자의 이름도 함께 가져오기

jungyoubin committed:
feat: 기존 next cursor를 통하여서 DB 조회하는 것에서 Offset으로 메시지 남은 개수 비교하는 기능으로 변경

jungyoubin committed:
feat: 채팅 페이지네이션 기능 수정

- DATE로 정렬을 할 경우, 중복이 발생할 수 있다.
- 중복이 발생할 경우, 데이터가 누락이 될 수 있기에 uuidv7을 통해 정렬 및 cursor 지정
- limit보다 1개를 더 확인하여서 뒤에 있는 값의 여부 확인하기

sungmin22 commented on src/chat/chat.service.ts:
```diff
@@ -300,18 +300,27 @@ export class ChatService {
-    if (rows.length === 0) {
-      // 메시지 없으면 종료
-      return { roomId, messages: [], hasMore: false };
+    const hasMore = rows.length > limit; // 뒤에 더 있는지 확인
+
+    // 응답으로 내보낼 실제 데이터 배열
+    // hasMore이 true면, 1개를 버리고 limit갯수만, false이면 rows 보내기
+    const page = hasMore ? rows.slice(0, limit) : rows;
+
+    if (page.length === 0) {
+      return { roomId, messages: [], nextCursor: null, hasMore: false };
     }
```
`hasMore`의 경우 로직 내에서만 지정하고 사용되는 값으로 보이는데, 반환되어야하는 이유가 혹시 있을까요?

jungyoubin replied:
사용자가 다음 페이지를 요구하려고 하였을때, `hasMore`을 통해서 뒤에 남아있는 데이터를 있는지에 대한 여부를 확인하기 위해 다음과 같이 설정하였습니다. 

만약에 휴대전화에서 스크롤을 통해서 이전 자료들이 받아와지는 형식이라면, 데이터가 뒤에 더 없지만 뒤에 더 있는줄 알고 계속해서 스크롤을 진행하려고 하였을때, 데이터가 없다는 문구를 알려주기 위해서 사용하였습니다.

sungmin22 replied:
`hasMore` 없이도 클라이언트에서 `nextCursor`가 `null`인지 여부를 통해서도 해당 페이지 뒤에 추가 데이터가 있는지 확인할 수 있을 것으로 보이는데, 의견 부탁드립니다!

jungyoubin replied:
`nextCursor`가 `null` 이 나오는 경우는 해당 채팅방에 아무러 메시지가 없는 경우일때 입니다. 초반에 어느정도의 데이터가 있고 데이터들을 계속해서 불러오고 있는 상황에서는 `nextCursor`가 맨 마지막의 데이터인 본인을 가리키는 경우 사용자의 입장에서는 `hasMore`가 없다면 더 있는지에 대한 여부를 확인하기 어렵다고 생각하여 다음과 같이 진행하였습니다.

sungmin22 replied:
`nextCursor`가 반환되는 데이터 목록의 가장 마지막 항목을 가리키고 있기 때문에 `hasMore`를 추가하셨다는 것 확인했습니다. 이 방식을 디벨롭 할 수 있을 것 같은데, 이는 리팩토링 요소로 두고 후에 함께 논의 해보면 좋을 것 같습니다.

sungmin22 reviewed (approved):
수정사항 모두 확인헀습니다. 수고 많으셨습니다.

Yuhan-Jeong commented on src/chat/chat.service.ts:
```diff
@@ -299,6 +299,56 @@ export class ChatService {
+      messageDate: new Date(r.messageDate).toISOString(), // ISO 문자열로 통일
+    }));
+
+    const nextCursor = messages[messages.length - 1].messageId;
```
`nextCursor`는 `page.length`가 0이 아닌 이상 항상 반환될 것으로 보입니다. 개선이 가능할 것 같은데 코드리뷰 시간에 확인해보시죠.

Yuhan-Jeong reviewed (approved):
수고하셨습니다. 아래 작성된 사항은 이후 코드리뷰시간 또는 추후에 리팩토링 요소로 확인만 부탁드립니다.
